### PR TITLE
ApiConsumer fixes #974

### LINKED ISSUE #974: RPC helpers for accessing 3rd party data
there are many RPCs  that only serve third party data with some processing on this data. A utility is needed to factor out the repetitive work from these rpcs. 
The utility would need support  sending the data in two formats:
1. as messages 
2. as a data structure. eg. tuples. 2d key, val arrays

### PR DESCRIPTION
**Pull request by hamidzr**
introducing an api consumer class to streamline the creation of RPCs

### CODE REVIEW

**Comment by hamidzr** [comment]
what do you think?

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+     */
+    _sendStruct(queryOptions,parserFn){
+        return this._requestData(queryOptions)
+            .then(res => {
+                this._logger.trace('got response back',res.length);
+                let parsedRes =  parserFn(res);
+                this._logger.trace(parsedRes);
+                // TODO check if parserFn is doing ok
+                try{
+                    if (Array.isArray(parsedRes)) {
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: be consistent with your array checking (you use a different method above)

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+            }
+        });
+        imgResponse.on('error', err => {
+            logger.error(err);
+            // QUESTION what to return in case of an error?
+            response.status(404).send('null');
+        });
+    }
+
+    // WIP
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: ?

**Comment by hamidzr** [comment]
are the exceptions generated by the services are absorbed by netsblox?

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+     * request a full response sending back a data structure.
+     * @param  {string} queryOptions
+     * @param  {Function} parseFn should convert a valid response from the api into an array of simple (tuple - keyvalue) Json Objects for  messages
+     * @return {Promise}
+     */
+    _sendStruct(queryOptions,parserFn){
+        return this._requestData(queryOptions)
+            .then(res => {
+                this._logger.trace('got response back',res);
+                let parsedRes =  parserFn(res);
```
Comment (hamidzr) [inline src/server/rpc/procedures/utils/api-consumer.js]: Do you think I should a try catch around this parser ? then when the user is writing the parserFn can throw errors when those happen.

**Comment by brollb** [comment]
Can you check out the TODO's caught by codeclimate?

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+        return imgResponse;
+    }
+
+    // private
+    _sendNext() {
+        var msgs = remainingMsgs[this.socket.uuid];
+        if (msgs && msgs.length) {
+            var msg = msgs.shift();
+
+            // while (msgs.length && msg.dstId !== this.socket.uuid) {
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: remove this if no longer needed

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+     */
+    _requestImage(queryOptions){
+        // TODO add optional caching
+        let fullUrl = (queryOptions.baseUrl || this._baseUrl) + queryOptions.queryString;
+        var imgResponse = request.get(fullUrl);
+        this._logger.trace('requesting image', fullUrl);
+        delete imgResponse.headers['cache-control'];
+        imgResponse.isImage = true;
+        imgResponse.on('response', res => {
+            if (!res.headers['content-type'].startsWith('image')) {
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: update this to use cache like from static map rpc

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+                response.set('content-type', 'image/png');
+                response.set('content-length', imageBuffer.length);
+                response.set('connection', 'close');
+                response.status(200).send(imageBuffer);
+                logger.trace('sent the image');
+            }
+        });
+        imgResponse.on('error', err => {
+            logger.error(err);
+            // QUESTION what to return in case of an error?
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: address or delete

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+            response.status(404).send('null');
+        });
+    }
+
+    // WIP needs further testing
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: probably just remove this and the following fn for now

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+    // = request data that logs then logs the response and limits it by selector
+    _inspectResponse(queryOptions, selector){
+        this._requestData(queryOptions)
+            .then(res=> {
+                this._logger.trace('got response');
+                this._logger.trace(this._queryJson(res,selector));
+            });
+    }
+
+    // QUESTION how to make this apply whenever sendMsgs() is used?
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: I don't think this is an issue. we can probably remove this comment as long as it all works

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+            }
+        } catch (e) {
+            this._logger.error('input is not valid json');
+        }
+        return jsonQuery(query, {data: json}).value;
+    }
+
+
+    // creates snap friendly structure out of an array ofsimple keyValue json object or just single on of them.
+    _createSnapStructure(jsonData){
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: This should be easy to add some tests for

File: src/server/rpc/procedures/utils/api-consumer.js [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+                setTimeout(this._sendNext.bind(this), MSG_SENDING_DELAY);
+            } else {
+                delete remainingMsgs[this.socket.uuid];
+            }
+        } else {
+            delete remainingMsgs[this.socket.uuid];
+        }
+    }
+
+    // TODO move this out of the class as a helper
```
Comment (brollb) [inline src/server/rpc/procedures/utils/api-consumer.js]: probably delete this comment

**Comment by hamidzr** [comment]
should be all good now.

**Pull request merged**